extensions/eclipse-che-theia-terminal/src/node/workspace-service-impl.ts

```diff
@@ -26,80 +26,70 @@ export class CHEWorkspaceServiceImpl implements CHEWorkspaceService {
     constructor(@inject(EnvVariablesServer) protected readonly baseEnvVariablesServer: EnvVariablesServer) {
     }
 
-    public async getMachineList(): Promise<{ [attrName: string]: che.workspace.Machine }> {
-        const machineNames: { [attrName: string]: che.workspace.Machine } = {};
-        const workspaceId = await this.getWorkspaceId();
-        const restClient = await this.getRemoteApi();
-        if (!workspaceId || !restClient) {
-            return machineNames;
-        }
-        return new Promise<{ [attrName: string]: che.workspace.Machine }>((resolve, reject) => {
-            restClient.getById<che.workspace.Workspace>(workspaceId)
-                .then((workspace: che.workspace.Workspace) => {
-                    if (workspace.runtime) {
-                        resolve(workspace.runtime.machines);
-                        return;
+    public async getContainerList(): Promise<WorkspaceContainer[]> {
+        const containers: WorkspaceContainer[] = [];
+        try {
+            const workspaceId = await this.getWorkspaceId();
+            const restClient = await this.getRemoteApi();
+            if (!workspaceId || !restClient) {
+                throw new Error('Unable to use workspace client.');
+            }
+
+            const workspace = await restClient.getById<che.workspace.Workspace>(workspaceId);
+
+            if (workspace.runtime && workspace.runtime.machines) {
+                const machines =  workspace.runtime.machines;
+                for (const machineName in machines) {
+                    if (!machines.hasOwnProperty(machineName)) {
+                        continue;
                     }
-                    resolve({});
-                })
-                .catch((reason: IRequestError) => {
-                    console.log(`Failed to get workspace by ID: ${workspaceId}. Status code: ${reason.status}`);
-                    reject(reason.message);
-                });
-        });
+                    const machine = workspace.runtime.machines[machineName];
+                    const container: WorkspaceContainer = {name: machineName, ...machine};
+                    containers.push(container);
+                }
+            }
```

Comment:
And generate an error if there are no runtime in the workspace config

Reply:
I guess this error never appears, because runtime is empty only if workspace is stopped. But in this case all theia and all extensions are stopped too.

Reply:
It is API part and errors can have the place. 
If you go thru this way - remove you ' if (workspace.runtime  ->  just use  'workspace!.runtime'